rust-src/bulletproofs/src/set_non_membership_proof.rs

```diff
@@ -0,0 +1,668 @@
+    tx:       C::Scalar,
+    /// Blinding factor for the commitment to `tx`
+    tx_tilde: C::Scalar,
+    /// Blinding factor for the commitment to the inner-product arguments
+    e_tilde:  C::Scalar,
+    /// Inner product proof
+    ip_proof: InnerProductProof<C>,
+}
+
+/// Error messages detailing why proof generation failed
+#[derive(Debug, PartialEq, Eq)]
+pub enum ProverError {
+    /// The length of the generator vector `gens` was too short
+    NotEnoughGenerators,
+    /// Could find the value `v` in the given set
+    CouldFindValueInSet,
+    /// Could not generate inner product proof
+    InnerProductProofFailure,
+    /// Could not invert an element
+    DivisionError,
+}
+
+/// This function produces a set-non-membership proof, i.e., a proof of
+/// knowledge of a value v that is not in a given set `the_set` and that is
+/// consistent with the commitment `V` to `v`. The arguments are
+/// - `transcript` - the random oracle for Fiat Shamir
+/// - `csprng` - cryptographic safe randomness generator
+/// - `the_set` - the set as a vector of scalars
+/// - `v` the value, a scalar
+/// - `gens` - generators containing vectors `G` and `H` both of at least length
+///   `k` where k is the smallest power of two >= `|the_set|`
+/// - `v_keys` - commitment keys `B` and `B_tilde` (`g,h` in the bluepaper)
+/// - `v_rand` - the randomness used to commit to `v` using `v_keys`
+#[allow(non_snake_case)]
+pub fn prove<C: Curve, R: Rng>(
+    transcript: &mut RandomOracle,
+    csprng: &mut R,
+    the_set: &[C::Scalar],
+    v: C::Scalar,
+    gens: &Generators<C>,
+    v_keys: &CommitmentKey<C>,
+    v_rand: &Randomness<C>,
+) -> Result<SetNonMembershipProof<C>, ProverError> {
+    // Part 0: Add public inputs to transcript
+    // Domain separation
+    transcript.add_bytes(b"SetNonMembershipProof");
+    // Compute commitment V for v
+    let v_value = Value::<C>::new(v);
+    let V = v_keys.hide(&v_value, v_rand);
+    // Append V to the transcript
+    transcript.append_message(b"V", &V.0);
+    // Pad set if not power of two
+    let mut set_vec = the_set.to_vec();
+    pad_vector_to_power_of_two(&mut set_vec);
+    let n = set_vec.len();
+    // Append the set to the transcript
+    transcript.append_message(b"theSet", &set_vec);
+
+    // Part 1: Setup and generation of vector commitments
+    // Check that we have enough generators for vector commitments
+    if gens.G_H.len() < n {
+        return Err(ProverError::NotEnoughGenerators);
+    }
+
+    // Get generator vector for blinded vector commitments, i.e. (G,H,B_tilde)
+    let mut GH_B_tilde: Vec<C> = Vec::with_capacity(2 * n + 1);
+    let (mut G, mut H): (Vec<_>, Vec<_>) = gens.G_H.iter().take(n).cloned().unzip();
+    GH_B_tilde.append(&mut G);
+    GH_B_tilde.append(&mut H);
+
+    // Generators for single commitments and blinding
+    let B = v_keys.g;
+    let B_tilde = v_keys.h;
+    GH_B_tilde.push(B_tilde);
+
+    // define aliases to make G and H available again
```

Comment:
I don't understand this comment. What does "available again" mean?

Reply:
`GH_B_tilde.append(&mut G);` above moves the elements out of G; this redefines it as the corresponding sub-slice. Probably that's not the right way to do it, though. ;) We tried some things and tested for performance, but generally the fact that `gens` contains `G` and `H` zipped does not help here...

Reply:
I see. Probably a more appropriate solution would be
```
let (G, H) = gens.G_H.iter().take(n).unzip();
GH_B_tilde.extend(G.chain(H).copied();
let G = &GH_B_tilde[0..n];
let H = &GH_B_tilde[n..2 * n];
```

Reply:
Rust doesn't let me unzip without cloning...

Reply:
Cool, we'll use that then. But yeah, it doesn't really matter, and we copy things multiple times in other places.